pkg/sources/elasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,270 @@
+
+				for _, docSearch := range uow.documentSearches {
+					documentsProcessed, err := processSearchedDocuments(
+						s.ctx,
+						client,
+						&docSearch,
+						func(document *Document) error {
+							if docSearch.index.DocumentAlreadySeen(document) {
+								return nil
+							}
+
+							chunk := sources.Chunk{
+								SourceType: s.Type(),
+								SourceName: s.name,
+								SourceID:   s.SourceID(),
+								JobID:      s.JobID(),
+								SourceMetadata: &source_metadatapb.MetaData{
+									Data: &source_metadatapb.MetaData_Elasticsearch{
+										Elasticsearch: &source_metadatapb.Elasticsearch{
+											Index:      sanitizer.UTF8(docSearch.index.name),
+											DocumentId: sanitizer.UTF8(document.id),
+											Timestamp:  sanitizer.UTF8(document.timestamp),
+										},
+									},
+								},
+								Verify: s.verify,
+							}
+
+							chunk.Data = []byte(document.message)
+
+							return common.CancellableWrite(ctx, chunksChan, &chunk)
+						},
+					)
+					if err != nil {
+						return err
+					}
+
+					fmt.Printf(
+						"[Worker %d] Scanned %d documents from index %s\n",
+						uowIndex,
+						documentsProcessed,
+						docSearch.index.name,
+					)
+
+					s.log.V(2).Info(fmt.Sprintf(
+						"[Worker %d] Scanned %d documents from index %s",
+						uowIndex,
+						documentsProcessed,
+						docSearch.index.name,
+					))
+
+					if documentsProcessed != docSearch.documentCount-docSearch.skipCount {
+						s.log.V(1).Info(fmt.Sprintf(
+							"documentsProcessed != docSearch.documentCount = docSearch.skipCount (%d != %d)",
+							documentsProcessed,
+							docSearch.documentCount-docSearch.skipCount,
+						))
+					}
+
+					uowDocumentsProcessed += documentsProcessed
+					indices.UpdateProcessedDocumentCount(documentsProcessed)
+					s.SetProgressComplete(
+						indices.GetProcessedDocumentCount(),
+						indices.documentCount,
+						fmt.Sprintf(
+							"[Worker %d] Scanned %d documents from index %s",
+							uowIndex,
+							documentsProcessed,
+							docSearch.index.name,
+						),
+						"",
+					)
+
+					// When we use the Elastic API in this way, we can't tell
+					// it to only return a specific number of documents. We can
+					// only say "return a page of documents after this offset".
+					// So we might have reached the limit of how many documents
+					// we're supposed to process with this worker in the middle
+					// of a page, so check for that here.
+					//
+					// (We could use the API in a different way to get a
+					// precise number of documents back, but that use is
+					// limited to 10000 documents which we could well exceed)
+					if uowDocumentsProcessed >= uow.documentCount {
+						break
+					}
+
+					docSearch.index.UpdateLatestTimestampLastRun()
+				}
+
+				return nil
+			})
+		}
+
+		err = workerPool.Wait()
+		if err != nil {
+			s.log.V(2).Info(fmt.Sprintf("Error waiting on worker pool: %s\n", err))
+		}
+
+		if !s.bestEffortScan {
```

Comment:
**Question**: In the case where we are doing continuous scans will this burn up the ES API? Will we run into rate limits? Without having run an ES scan, my intuition suggests we put some sort of sleep after each iteration.

Reply:
Great point; yeah will add a `Sleep`